File: php/src/DynamicKey4.php
<?php

    function generateRecordingKey($staticKey, $signKey, $channelName, $ts, $randomInt, $uid, $expiredTs ,$serviceType='ARS')
    {
        return generateDynamicKey($staticKey, $signKey, $channelName, $ts, $randomInt, $uid, $expiredTs ,$serviceType);
    }

    function generateMediaChannelKey($staticKey, $signKey, $channelName, $ts, $randomInt, $uid, $expiredTs ,$serviceType='ACS')
    {
        return generateDynamicKey($staticKey, $signKey, $channelName, $ts, $randomInt, $uid, $expiredTs ,$serviceType);
    }

    function generateDynamicKey($staticKey, $signKey, $channelName, $ts, $randomInt, $uid, $expiredTs ,$serviceType)
    {
        $version = "004";

        $randomStr = "00000000" . dechex($randomInt);
        $randomStr = substr($randomStr,-8);

        $uidStr = "0000000000" . $uid;
	    $uidStr = substr($uidStr,-10);
        
        $expiredStr = "0000000000" . $expiredTs;
        $expiredStr = substr($expiredStr,-10);

        $signature = generateSignature($staticKey, $signKey, $channelName, $ts, $randomStr, $uidStr, $expiredStr ,$serviceType);

        return $version . $signature . $staticKey . $ts . $randomStr . $expiredStr;
    }

    function generateSignature($staticKey, $signKey, $channelName, $ts, $randomStr, $uidStr, $expiredStr ,$serviceType)
    {
        $concat = $serviceType . $staticKey . $ts . $randomStr . $channelName . $uidStr . $expiredStr;
        return hash_hmac('sha1', $concat, $signKey);
    }
?>

File: go/src/DynamicKey4/DynamicKey4.go
package DynamicKey4

import (
	"crypto/hmac"
	"crypto/sha1"
	"encoding/hex"
	"fmt"
	"strings"
)

func GeneratePublicSharingKey(vendorKey, signKey, channelName string, unixTs, randomInt, uid, expiredTs uint32) (string) {
    return generateDynamicKey(vendorKey, signKey, channelName, unixTs, randomInt, uid, expiredTs, "APSS")
}

func GenerateRecordingKey(vendorKey, signKey, channelName string, unixTs, randomInt, uid, expiredTs uint32) (string) {
    return generateDynamicKey(vendorKey, signKey, channelName, unixTs, randomInt, uid, expiredTs, "ARS")
}

func GenerateMediaChannelKey(vendorKey, signKey, channelName string, unixTs, randomInt, uid, expiredTs uint32) (string) {
    return generateDynamicKey(vendorKey, signKey, channelName, unixTs, randomInt, uid, expiredTs, "ACS")
}

func generateDynamicKey(vendorKey, signKey, channelName string, unixTs, randomInt, uid, expiredTs uint32, serviceType string) (string) {
    version := "004"
	unixTsStr := fmt.Sprintf("%010d", unixTs)
	randomIntStr := fmt.Sprintf("%08x", randomInt)
	uidStr := fmt.Sprintf("%010d", uid)
	expiredTsStr := fmt.Sprintf("%010d", expiredTs)
	signature := generateSignature(vendorKey, signKey, channelName, unixTsStr, randomIntStr, uidStr, expiredTsStr, serviceType)
	buffer := strings.Join([]string{version, signature, vendorKey, unixTsStr, randomIntStr, expiredTsStr}, "")
	return buffer
}

func generateSignature(vendorKey, signKey, channelName, unixTsStr, randomIntStr, uidStr, expiredTsStr, serviceType string) (string) {
	buffer := strings.Join([]string{serviceType, vendorKey, unixTsStr, randomIntStr, channelName, uidStr, expiredTsStr}, "")
	signature := hmac.New(sha1.New, []byte(signKey))
	signature.Write([]byte(buffer))
	return hex.EncodeToString(signature.Sum(nil))
}
